### DISCUSSION TITLE
gradient question about GatherScatterMode

### DISCUSSION THREAD

**Discussion by Dong-Jiahuan** [jax-ml/jax#17675] [Q&A] [answered]
Hello, In https://jax.readthedocs.io/en/latest/jax.lax.html#jax.lax.GatherScatterMode, it says ``Gradients will not be correct if indices are out-of-bounds. (in PROMISE_IN_BOUNDS mode)``

(1) I want to know If gradients will be correct or not when indices are out-of-bounds in modes other than PROMISE_IN_BOUNDS ?
(2) For example, the document says ``out-of-bounds gathers will be clamped``. Does it mean when you use lax.gather and indices are out-of-bounds and choose ``PROMISE_IN_BOUNDS`` mode, it's totally equivalent to choose ``CLIP`` mode ? (the results are same, gradients are both incorrect in both modes...)

These two questions above are the same question to some degree...

**Comment by hawkinsp** [comment, marked as answer]
1) Only `PROMISE_IN_BOUNDS` mode has incorrect gradients. The other modes will have correct gradients.
2) Exactly what `PROMISE_IN_BOUNDS` does is up to the implementation, but it's is equivalent to `CLIP` in the forwards direction, while its gradient is equivalent to `DROP`. (These don't match, which is why the result isn't correct.)

**Reply by Dong-Jiahuan** [reply]
I see. Thanks!
